### PR TITLE
Revert "singlehtml: deprecate the 'fix_refuris' helper function (#13037)"

### LINKED ISSUE #14221: Incorrect HTML anchors generated by singlehtml builder
### Describe the bug

Using the singlehtml builder to generate index.html from the example below produces invalid internal HTML anchors having multiple `#` characters in them with certain combinations of glossaries and toctree inclusions. In the reproducer below, there is an `index.rst` file with a toctree that includes two glossary files in separate documents, each with their own glossary item definition. An equivalent behavior can be observed also by using `.. include` to include the files directly. The build command is `sphinx-build -b singlehtml . _build/foo` and the HTML file is tested with `html5validator --root _build/foo --show-warnings`. Issue #13037 might be related.

### How to Reproduce

conf.py:
```
project = 'Glossary Test'
extensions = []
```

index.rst:
```
Main Documentation
==================

Welcome to the documentation.

.. toctree::
   :maxdepth: 2

   ch1
   glossary/term1
   glossary/term2
```
with the glossary files being trivial items:
```
.. glossary::

   API
      Lorem Ipsum.
```
The generate document contains:
```
<a class="reference internal" href="#document-glossary/term1#term-API"><span class="xref std std-term">API</span></a>
```

### Environment Information

```text
Platform:              linux; (Linux-6.14.0-37-generic-x86_64-with-glibc2.39)
Python version:        3.12.3 (main, Nov  6 2025, 13:44:16) [GCC 13.3.0])
Python implementation: CPython
Sphinx version:        9.0.4
Docutils version:      0.21.2
Jinja2 version:        3.1.4
Pygments version:      2.19.2
```

### Sphinx extensions

```python
None
```

### Additional context

Some build warnings are generated, but they should not affect the outcome of the test

```
writing... /home/nicola/test-sphinx/index.rst:6: WARNING: toctree contains reference to document 'glossary/term1' that doesn't have a title: no link will be generated [toc.no_title]
/home/nicola/test-sphinx/index.rst:6: WARNING: toctree contains reference to document 'glossary/term2' that doesn't have a title: no link will be generated [toc.no_title]
/home/nicola/test-sphinx/index.rst:6: WARNING: toctree contains reference to document 'glossary/term1' that doesn't have a title: no link will be generated [toc.no_title]
/home/nicola/test-sphinx/index.rst:6: WARNING: toctree contains reference to document 'glossary/term2' that doesn't have a title: no link will be generated [toc.no_title]
```

### PR DESCRIPTION
**Pull request by jayaddison**
## Purpose

As reported in #14221, the `fix_refuris` method does appear to prevent duplicate fragment delimiters (the `#` character) from appearing in the hyperlinks output by the `singlehtml` builder.

In my opinion we should still consider `fix_refuris` deprecated -- I'd still like to remove it, since it seems like an unnecessary post-processing of the documents; but additional investigation is required to confirm what it does and whether a better alternative mechanism can be found.

## References

- Accompanies #14240.
- Resolves #14221.
- Reverts #13037

### CODE REVIEW

**Comment by jayaddison** [comment]
Self-quoting:
> In my opinion we should still consider `fix_refuris` deprecated -- I'd still like to remove it, since it seems like an unnecessary post-processing of the documents; but additional investigation is required to confirm what it does and whether a better alternative mechanism can be found.

Please refer to this comment for the results of this investigation: https://github.com/sphinx-doc/sphinx/issues/14221#issuecomment-3710067954

**Comment by nvetrini** [comment]
Hi,

just checking if there is any feedback on this PR.

Thanks

**Comment by kreuzberger** [comment]
It would be helpful to fix the singlethml builds for sphinx 9.x therefore :+1: for this PR.
But as long as the checks are failing i think nobody will merge this?

**Comment by nvetrini** [comment]
Hi @kreuzberger . Thank you for the feedback. I might try to fix those if I have time.

For the Python 3.12 check only one failed test, as far as I can see. It has been reported to be unreliable in #14104

**Comment by jayaddison** [comment]
This depends on #14240, that is ready for review.

File: sphinx/builders/singlehtml.py [hunk 1]
```diff
@@ -53,7 +53,7 @@ def get_relative_uri(self, from_: str, to: str, typ: str | None = None) -> str:
 
     def fix_refuris(self, tree: Node) -> None:
         deprecation_msg = (
-            "The 'SingleFileHTMLBuilder.fix_refuris' method is no longer used "
+            "The 'SingleFileHTMLBuilder.fix_refuris' method is deprecated "
```
Comment (jdillard) [inline sphinx/builders/singlehtml.py]: Since `fix_refuris()` emits `RemovedInSphinx10Warning` and Sphinx enables these
warnings by default, restoring these internal calls would cause an
ordinary `singlehtml` build to emit deprecation warnings from Sphinx's own
internal use.

This would be new behavior for a released Sphinx version:

- Through Sphinx 8.1.x the internal calls existed, but the method did not warn
- From 8.2.0 onward the method warns but is not called internally within Sphinx
- This PR combines internal calls with the warning for the first time, so normal use of the `singlehtml` builder would emit deprecation warnings

The deprecation warning is designed to help extension authors who call the deprecated public API prepare for its removal. Would it therefore make sense for the reference-rewriting logic to move to a non-warning private helper that Sphinx calls internally, with the public `fix_refuris()` method emitting the deprecation warning and then delegating to the private helper?

```python
def _fix_refuris(self, tree):
    # implementation

def fix_refuris(self, tree):
    warnings.warn(...)
    self._fix_refuris(tree)
```

That would preserve an actionable deprecation for external callers and allow the public method to be removed in Sphinx 10 while retaining the internal behavior that this regression shows is still needed. Later, we can replace and remove the private post-processing helper when a better URI-generation design is implemented.

File: sphinx/builders/singlehtml.py [hunk 2]
```diff
@@ -86,6 +86,8 @@ def _get_local_toctree(
         toctree = global_toctree_for_doc(
             self.env, docname, self, tags=self.tags, collapse=collapse, **kwargs
         )
+        if toctree is not None:
+            self.fix_refuris(toctree)
```
Comment (jdillard) [inline sphinx/builders/singlehtml.py:80]: Not related to this line, but should there be a `CHANGES.rst` entry under the Sphinx 9.1.1 bug fixes for
#14221? This is a user-visible regression fix affecting releases from 8.2.0 onward.

File: sphinx/builders/singlehtml.py (hunk 1 above)
Comment (jayaddison) [reply sphinx/builders/singlehtml.py]: Thanks @jdillard - yes, we should make sure that running the `singlehtml` builder does not emit this deprecation warning.

Sphinx 10 might not be too far in the future at this point, and I am not certain I'll find time to identify and/or refactor the code that requires the `fix_refuris` patching.  So I wonder whether we should remove the deprecation warning entirely, and re-attempt deprecation/removal again in future?

Comment (jdillard) [reply sphinx/builders/singlehtml.py]: Yes, I think withdrawing the deprecation is cleaner. Once the replacement exists, we can begin a new deprecation cycle.

Comment (jayaddison) [reply sphinx/builders/singlehtml.py]: Ok; I've removed the deprecation (after one false start, because I'd forgotten that there is test coverage for it..).

File: tests/test_builders/test_build_html_refuris.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+def _internal_reference_fragment_check(nodes: Sequence[Element]) -> None:
+    """Confirm that internal references do not contain duplicate fragment symbols"""
+    assert nodes, 'Expected at least one node to check'
+    for node in nodes:
+        assert node.tag == 'a', 'Attempted to check hyperlink on a non-anchor element'
+        href = node.attrib.get('href')
+        # Allow Sphinx index and table hyperlinks to be non-same-document, as exceptions.
+        if href in {'genindex.html', 'py-modindex.html', 'search.html'}:
+            continue
+        assert not href or href.count('#') < 2, 'Hyperlink contains duplicate fragments'
```
Comment (jdillard) [inline tests/test_builders/test_build_html_refuris.py]: This assertion would also pass for a single-fragment link pointing to the wrong target.

I think it would be better to assert that the glossary references resolve specifically to `#term-API`, and that the target exists. That would verify that the destination is preserved rather than only checking that one `#` was removed.

Replacing `test_singlehtml_refuris` with this should accomplish that:

```python
@pytest.mark.sphinx('singlehtml', testroot='refuris')
def test_singlehtml_refuris(
    app: SphinxTestApp,
    cached_etree_parse: Callable[[Path], ElementTree],
) -> None:
    app.build()
    etree = cached_etree_parse(app.outdir / 'index.html')
    api_refs = [
        node
        for node in etree.findall(".//a[@class='reference internal']")
        if ''.join(node.itertext()) == 'API'
    ]
    assert api_refs
    assert all(ref.get('href') == '#term-API' for ref in api_refs)
    assert etree.find(".//*[@id='term-API']") is not None
```

The `_internal_reference_fragment_check` helper can then be removed, along with the now-unused `check_xpath`, `Sequence`, and `Element` imports.

File: tests/roots/test-refuris/ch1.rst [hunk 4]
```diff
@@ -0,0 +1,14 @@
+Chapter One
+===========
+
+This chapter discusses :term:`API` design and :term:`REST` principles.
```
Comment (jdillard) [inline tests/roots/test-refuris/ch1.rst]: The regression only requires one term reference in a different document from
its definition.

To make the test’s purpose clearer, we could use just a single `API` term reference in `ch1.rst` and a single `API` definition in `glossary/term1.rst`, removing both the `REST` references and `glossary/term2.rst` (along with its toctree entry). This would also eliminate the unrelated warnings currently produced. Adding a title to `term1.rst` would eliminate the no-title warning as well.

File: tests/test_builders/test_build_html_refuris.py (hunk 3 above)
Comment (jayaddison) [reply tests/test_builders/test_build_html_refuris.py]: @jdillard I take your point, and your suggested test coverage seems reasonable to me (I've run it against the `main` branch locally to confirm that it fails).

Thinking back to the intent of my test coverage, though, it was to express "no hyperlinks on the page should contain more than one fragment identifier" -- a general statement about the emitted HTML, rather than about the specific `API` term references.

Could/should we include both?  (perhaps as separate test cases?)

Comment (jdillard) [reply tests/test_builders/test_build_html_refuris.py]: Got it. I agree and think we should include both as separate tests.

File: tests/roots/test-refuris/ch1.rst (hunk 4 above)
Comment (jayaddison) [reply tests/roots/test-refuris/ch1.rst]: I agree with reducing the definitions to a single `term` :+1: 

About the removal of the `toctree` from `index.rst`: I don't think that the bug will exhibit without that (or, at least, we would need to restructure the test project further to make it appear again).

Comment (jdillard) [reply tests/roots/test-refuris/ch1.rst]: I just wanted `glossary/term2.rst` removed, I don't think it requires a duplicate entry for this regression.

Comment (jayaddison) [reply tests/roots/test-refuris/ch1.rst]: You're correct - I'd misunderstood that (and your review/change request here).

**Comment by jayaddison** [comment]
The latest unit test failures (e.g. on cf5af53f82f7c12817508f5facf5825537e5feee) seem to relate to HTML encoding differences - single-quotes and double-quotes are no-longer HTMLescaped by Sphinx (to `&#39;` and `&quot;` respectively), for example, while the test suite continues to expect that escaping in a few places.

I'm not sure what has caused that behaviour to change, though.  Perhaps the Python version used by the GitHub runners has updated with a security fix?

I'll report a CI failure issue, and have some local changes prepared that should fix it -- but it would be good to figure out why it happened.

File: tests/roots/test-refuris/glossary/term1.rst [hunk 5]
```diff
@@ -0,0 +1,4 @@
+.. glossary::
```
Comment (jdillard) [inline tests/roots/test-refuris/glossary/term1.rst:4]: Fix the a `[toc.no_title]` warning, just to keep things clean:

```suggestion
Glossary
========

.. glossary::
```

File: tests/test_builders/test_build_html_refuris.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
```
Comment (jdillard) [inline tests/test_builders/test_build_html_refuris.py]: The catch for these three special-case filenames can be removed, they don't  contain any `#`, so they already pass the assertion `href.count('#') < 2`.

The comment can also be removed as it suggests this test rejects non-same-document links, but it only checks for duplicate fragment delimiters.

```suggestion
        assert not href or href.count('#') < 2, 'Hyperlink contains duplicate fragments'
```

Comment (jayaddison) [reply tests/test_builders/test_build_html_refuris.py]: That makes sense, yep.  I'm going to push one more commit/refactor along with that, to remove the `not href or` part of the assertion; it seems inelegant.

Comment (jayaddison) [reply tests/test_builders/test_build_html_refuris.py]: I couldn't remember why I added that comment, so I went back to re-run the test from the commit where I added it, but without the fix in place; it didn't have an effect.

It seems I just copy-pasted it from a similar node assertion helper method I'd written, and didn't audit it for removal: https://github.com/sphinx-doc/sphinx/blob/e44a40eb2f810558ccd9da1425421270ccb81351/tests/test_builders/xpath_html_util.py#L16-L18

File: CHANGES.rst [hunk 7]
```diff
@@ -13,6 +13,11 @@ Bugs fixed
   English stemmer) and Dutch (which uses the Dutch Porter stemmer).
   Patch by Hugo van Kemenade
 
+* #14221: Fix a regression from Sphinx v8.2.0 onwards that could cause
+  duplicate hyperlink fragment identifiers to be emitted during ``singlehtml``
+  project builds.
+  Patch by James Addison
+
```
Comment (jdillard) [inline CHANGES.rst]: Could this mention both the malformed URI and the withdrawal of the `SingleFileHTMLBuilder.fix_refuris` deprecation? Technically, removing the warning and the entry from `deprecated.rst` is a user-visible API change for extension authors.

Also, the regression produces multiple fragment delimiters (`#`) in one URI, rather than duplicate fragment identifiers.

```suggestion
* #14221: Fix a regression from Sphinx v8.2.0 onwards that could cause links
  produced by ``singlehtml`` builds to contain multiple fragment delimiters
  (``#``), and withdraw the deprecation of
  ``SingleFileHTMLBuilder.fix_refuris`` because the builder still requires it.
  Patch by James Addison
```

**Review by jdillard — COMMENTED** [review]
This _should_ be the last review comment, I don't think I see anything else in the code.

Related: The `Purpose` section of the PR description is now out of sync with the implementation and says that `fix_refuris` should remain deprecated.

File: CHANGES.rst (hunk 7 above)
Comment (jayaddison) [reply CHANGES.rst]: Thanks for the re-review @jdillard - and yep, that edit seems good to me :+1:

**Review by jdillard — APPROVED** [review]
Thanks!
